src/meep.hpp

```diff
@@ -711,6 +713,80 @@ boundary_region pml(double thickness, double Rasymptotic = 1e-15, double mean_st
 
 class binary_partition;
 
+enum in_or_out { Incoming = 0, Outgoing };
+constexpr std::initializer_list<in_or_out> all_in_or_out{Incoming, Outgoing};
+
+enum connect_phase { CONNECT_PHASE = 0, CONNECT_NEGATE = 1, CONNECT_COPY = 2 };
+constexpr std::initializer_list<connect_phase> all_connect_phases{
+     CONNECT_PHASE, CONNECT_NEGATE, CONNECT_COPY};
+constexpr int NUM_CONNECT_PHASE_TYPES = 3;
+
+// Communication pair(i, j) implies data being sent from i to j.
+using chunk_pair = std::pair<int, int>;
+
+// Key for the map that stored communication sizes.
+struct comms_key {
+  field_type ft;
+  connect_phase phase;
+  chunk_pair pair;
+};
+
+// Defined in fields.cpp
+bool operator==(const comms_key &lhs, const comms_key &rhs);
+
+class comms_key_hash_fn {
+
+public:
+  inline std::size_t operator()(const comms_key &key) const {
+    // Unroll hash combination to promote the generatiion of efficient code.
+    std::size_t ret = ft_hasher(key.ft);
+    ret ^= connect_phase_hasher(key.phase) + kHashAddConst + (ret << 6) + (ret >> 2);
+    ret ^= int_hasher(key.pair.first) + kHashAddConst + (ret << 6) + (ret >> 2);
+    ret ^= int_hasher(key.pair.second) + kHashAddConst + (ret << 6) + (ret >> 2);
+    return ret;
+  }
+
+private:
+  static constexpr size_t kHashAddConst = 0x9e3779b9;
```

Comment:
Should this be a 64-bit value (i.e. not have 32 zero bits) on 64-bit machines?

I guess it doesn't matter for us since our hash maps shouldn't be getting anywhere near 2^31 entries.